### PR TITLE
Begin implementing Payment Receiver

### PR DESCRIPTION
**Pull request by Lewington-pitsos**
At this stage all we implement is the clientId database aspect of Payment Receiver.

We also clean up some un-needed code and add tests.

### CODE REVIEW

File: contracts/PaymentReceiver.sol [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    uint private nonce;
+    mapping (uint256 => address) private clientIds;
+    uint256[] private idList;
+    // An address cannot have more than 128 associated clients.
+    uint32 private maxClientIds = 128;
+
+    constructor() {}
+
+    function addClient() external returns (uint256) {
+        uint256 _clientId = uint256(keccak256(abi.encode(block.number, msg.data, nonce++)));
```
Comment (2manslkh) [inline contracts/PaymentReceiver.sol]: Any reason you would use a random clientID as opposed to an incremental? This costs extra gas. I would suggest the usage of the Counter class

File: contracts/IPaymentReceiver.sol [hunk 2]
```diff
@@ -1,5 +1,5 @@
 //SPDX-License-Identifier: MIT
-pragma solidity ^0.8.11;
+pragma solidity ^0.8.0;
```
Comment (2manslkh) [inline contracts/IPaymentReceiver.sol]: Lets keep to solidity 0.8.11

File: contracts/PaymentReceiver.sol [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+import "hardhat/console.sol";
+import "./IPaymentReceiver.sol";
+
+contract ClientDatabase {
+    uint private nonce;
+    mapping (uint256 => address) private clientIds;
+    uint256[] private idList;
+    // An address cannot have more than 128 associated clients.
+    uint32 private maxClientIds = 128;
```
Comment (2manslkh) [inline contracts/PaymentReceiver.sol:14]: What is the reason?

**Review by 2manslkh — CHANGES_REQUESTED** [review]
We could use typescript if you are familiar, but not necessary. Thanks for including the test cases too

File: contracts/PaymentReceiver.sol (hunk 3 above)
Comment (Lewington-pitsos) [reply contracts/PaymentReceiver.sol:14]: we need to choose the size of the array that we put client ids into as we iterate over it (apparently dynamic arrays in memory are not allowed)

File: contracts/PaymentReceiver.sol [hunk 4]
```diff
@@ -4,7 +4,7 @@ import "hardhat/console.sol";
 import "./IPaymentReceiver.sol";
 
 contract ClientDatabase {
-    uint private nonce;
+    uint private nextId;
```
Comment (2manslkh) [inline contracts/PaymentReceiver.sol]: https://docs.openzeppelin.com/contracts/4.x/api/utils#Counters

You might want to consider using Counters as a way to restrict _id from being increased more than 1

**Review by 2manslkh — CHANGES_REQUESTED** [review]

**Review by 2manslkh — APPROVED** [review]

**Pull request merged**